parser.go

```diff
@@ -555,27 +555,84 @@ var (
 	nullLiteral  = []byte("null")
 )
 
-func createInsertComponent(keys []string, setValue []byte, comma, object bool) []byte {
+func isValidArrayIndex(key string) (isArray bool, idx int) {
+	idx = -1
+	if key[0] == '[' && key[len(key)-1] == ']' {
+		idxVal := key[1 : len(key)-1]
+		if idxVal == "+" || idxVal == "-" {
+			isArray = true
+			idx = 0
+		} else if idxNum, err := strconv.Atoi(idxVal); err == nil {
+			if idxNum >= 0 {
+				isArray = true
+				idx = idxNum
+			}
+		}
+	}
+	return isArray, idx
+}
+
+func createInsertComponent(keys []string, setValue []byte, startComma, endComma, isObject bool) []byte {
```

Comment:
Could you add a couple comments to this (i.e. a top-level "what does this function do?" comment) and single-line comments on the major blocks?